design-proposals/klip-50-partition-and-offset-in-ksqldb.md

```diff
@@ -0,0 +1,253 @@
+transient queries, persistent queries, and also use the metadata in filters and UDFs.
+(See ["Public APIS"](#public-apis) below for example syntax.)
+
+This will allow users to more easily inspect data and debug ksqlDB queries:
+```
+SELECT *, ROWPARTITION, ROWOFFSET FROM my_stream EMIT CHANGES;
+```
+experiment with processing only a subset of messages:
+```
+SELECT * FROM my_stream WHERE ROWPARTITION = 0 AND ROWOFFSET > 12 EMIT CHANGES;
+```
+and skip select messages ("poison pills") during processing:
+```
+CREATE STREAM my_cleaned_stream AS SELECT * FROM my_stream WHERE NOT (ROWPARTITION = 0 AND ROWOFFSET = 42);
+```
+
+## Public APIS
+
+We will add two new [pseudocolumns](https://github.com/confluentinc/ksql/blob/master/design-proposals/klip-14-rowtime-as-pseudocolumn.md)
+to ksqlDB: `ROWPARTITION` (integer) and `ROWOFFSET` (bigint). We choose these names to be 
+consistent with the existing `ROWTIME` pseudocolumn, and to minimize the chance of conflict 
+with users' existing data columns for compatibility reasons. See 
+["Alternatives Considered"](#alternatives-considered) for discussion on possible alternatives.
+
+### Query Semantics
+
+The behavior of `ROWPARTITION` and `ROWOFFSET` in persistent queries and push queries
+will be analogous to those of the [existing `ROWTIME` pseudocolumn](https://github.com/confluentinc/ksql/blob/master/design-proposals/klip-14-rowtime-as-pseudocolumn.md#queries).
+In summary:
+
+* `SELECT *` projections will not include `ROWPARTITION` and `ROWOFFSET`
+* Users may choose to explicitly add `ROWPARTITION` or `ROWOFFSET` in the selection,
+  which will cause ksqlDB to copy the selected pseudocolumn(s) into the output _value_
+* `ROWPARTITION` and `ROWOFFSET` will become reserved column names, and users will not
+  be allowed to use these names either as source column names or as output column aliases.
+  (Existing persistent queries that already use these names will continue to run undisrupted.
+  See ["Compatibility Implications"](#compatibility-implications) for more.)
+
+Consistent with `ROWTIME`, the new `ROWPARTITION` and `ROWOFFSET` pseudocolumns will 
+also be allowed in `WHERE`, `GROUP BY`, and `PARTITION BY` clauses, though use cases
+for the latter two are probably limited. `ROWPARTITION` and `ROWOFFSET` will also be
+allowed in UDFs.
+
+### Pull Queries
+
+Today, users are able to select the `ROWTIME` pseudocolumn as part of pull queries:
+```
+SELECT *, ROWTIME FROM my_table WHERE my_key='foo';
+```
+For multiple reasons, we propose that the new `ROWPARTITION` and `ROWOFFSET` pseudocolumns
+should diverge from this behavior. In other words, users will NOT be allowed to select
+`ROWPARTITION` or `ROWOFFSET` as part of a pull query.
+
+The first reason is that selecting `ROWPARTITION` or `ROWOFFSET` for a table
+is semantically dubious, especially for tables that are the result of other
+computation such as an aggregation over a stream. Should `ROWPARTITION` and `ROWOFFSET`
```

Comment:
This argument may apply to persistent CSAS queries as well: should the partition/offset come from the resulted stream, or from the source stream record that generated the result record. Personally I feel just stick to say "it is always the resulted stream/table" is fine, but that also means we do not provide any way still for users to check the partition/offset of the source topics.

Reply:
Just commented the same thing, but I felt that it should come from the source not the result topic partition/offset 😆 - cc @colinhicks for some product insight 🔮

Reply:
I also agree they should be relative to the source.